Rough in Display Strings (#2494)
httpwg/http-extensions: draft-ietf-httpbis-sfbis.md

Strings are zero or more printable ASCII {{!RFC0020}} characters (i.e., the range %x20 to %x7E). Note that this excludes tabs, newlines, carriage returns, etc.

> Non-ASCII characters are not directly supported in Strings, because it causes a number of interoperability issues, and -- with few exceptions -- field values do not require it.

Author: ```suggestion
Non-ASCII characters are not directly supported in Strings, because they cause a number of interoperability issues, and -- with few exceptions -- field values do not require them.
```

> Unicode is not directly supported in Strings, because it causes a number of interoperability issues, and -- with few exceptions -- field values do not require it.

Reviewer: ```suggestion
Non-ASCII characters are not directly supported in Strings, because it causes a number of interoperability issues, and -- with few exceptions -- field values do not require it.
```

Example-Date: @1659578233
~~~

### Display Strings {#displaystring}

> Display Strings are similar to Strings, in that they consist of zero or more characters, but they allow unicode content, unlike Strings.

Reviewer 2: ```suggestion
Display Strings are similar to Strings, in that they consist of zero or more characters, but the content of a Display String is UTF-8-encoded Unicode, rather than the ASCII carried by a String.
```

Author: No - at this level (the data model), they're unicode strings; utf-8 is a characteristic of how they're serialised on the wire.

Reviewer 2: Oh, I see.  This is fine.  Capitalize "Unicode".

> Display Strings are intended for use in cases where a value is displayed to end users, and therefore may need to carry non-ASCII content. It is NOT RECOMMENDED that they be used in situations where a String ({{string}}) or Token ({{token}}) would be adequate, because unicode has processing considerations (e.g., normalization) and security considerations (e.g., homograph attacks) that make it more difficult to handle correctly.

Reviewer: ```suggestion
Display Strings are intended for use in cases where a value is displayed to end users, and therefore may need to carry non-ASCII content. It is NOT RECOMMENDED that they be used in situations where a String ({{string}}) or Token ({{token}}) would be adequate, because Unicode has processing considerations (e.g., normalization) and security considerations (e.g., homograph attacks) that make it more difficult to handle correctly.
```

Note that Display Strings do not indicate the language used in the value; that can be done separately if necessary (e.g., with a parameter).

For example:

~~~ http-message-new
Example-DisplayString: %"This is intended for display to %C3%BCsers."
~~~

> 

Reviewer: The prose should say what the display string in the example is

2. Return output.

### Serializing a Display String {#ser-display}

Given a string of Unicode characters as input_string, return an ASCII string suitable for use in an HTTP field value.

0. Let byte_array be the result of applying UTF-8 encoding {{UTF8}} to input_string. If there is an error in doing so, fail parsing.
1. Let encoded_string be an empty string.
2. For each byte in byte_array:
  1. If byte is %x25 ("%"), append "%25" to encoded_string.

>   2. If byte is in the ranges x00-1f or %x7f-ff, append the result of applying the percent-encoding defined in {{Section 2.1 of URI}} to encoded_string.

Author: How so? %x25 isn't in those ranges.
